webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/ncf/TechniqueReader.scala

```diff
@@ -101,20 +106,31 @@ class TechniqueReader(
       needsUpdate
     }
   }
-
-  def readMethodsMetadataFile : IOResult[Map[BundleName, GenericMethod]] = {
+  private[this] val methodsCache = Ref.make((Instant.EPOCH, Map[BundleName, GenericMethod]())).runNow
+  def getMethodsMetadata : IOResult[Map[BundleName, GenericMethod]] = {
     for {
-      _                    <- ZIO.whenM(doesMethodsMetadataFileNeedsUpdate) {
-                                updateMethodsMetadataFile.map(_.code).unit
-                              }
-      genericMethodContent <- IOResult.effect(s"error while reading ${methodsFile.pathAsString}")(methodsFile.contentAsString)
-      methods              <- parse(genericMethodContent) match {
+      methodsFileModifiedTime <- IOResult.effect(methodsFile.lastModifiedTime())
+      cache <- methodsCache.get
+      methods <- if (methodsFileModifiedTime.isAfter(cache._1)) {
+        readMethodsMetadataFile
+      } else {
+        cache._2.succeed
+      }
+    } yield {
+      methods
+    }
+  }
+def readMethodsMetadataFile : IOResult[Map[BundleName, GenericMethod]] = {
+   for {
+     genericMethodContent <- IOResult.effect(s"error while reading ${methodsFile.pathAsString}")(methodsFile.contentAsString)
+     methods              <- parse(genericMethodContent) match {
                                 case JObject(fields) =>
                                   restExtractor.extractGenericMethod(JArray(fields.map(_.value))).map(_.map(m => (m.id, m)).toMap).toIO
                                   .chainError(s"An Error occured while extracting data from generic methods ncf API")
 
                                 case a => Inconsistency(s"Could not extract methods from ncf api, expecting an object got: ${a}").fail
                               }
+     _ <- methodsCache.set((Instant.now(), methods))
```

Comment:
You current cache read/update is not atomic, you need to use one of the provided methods of `Ref` for that, see https://github.com/VinceMacBuche/rudder/pull/22

Reply:
(and you're right, cache manipulation needs to be done in `readMethodsMetadataFile` since it's use elsewhere)